### ISSUE TITLE
Trait based DependencyInjection

### ISSUE THREAD

**Issue by rijnhard** [symfony/symfony#13550]
I'd like to share a solution I had a while ago, I don't expect it to be incorporated. 

At a previous company I worked for we had a use case where we had our own cms, and for each plugin it needed particular services injected. 

When we had to declare these services in the yaml there was a looot of boilerplate, as you can imagine.

So the concept was that using a trait with a basic getter setter (some then turned out to do some pretty interesting things inside the setter) to define dependencies without the need for specifying calls or listing many services in the class constructors inside the services yaml.

``` php
trait LogTrait
{
    private $logger;

    public function setLogger(LoggerInterface $logger) ....
    public function getLogger() ....
}
```

To implement it only took a few compiler passes and a custom annotation, but there are many options. 
# Flow
1. compiler pass collects traits, and sets placeholders for dependencies to be injected into the class using the trait.
   - this can be done with an annotation or with config most likely, which would look similar to how services are defined.
2. second compiler pass uses the traits that were collected (depending on annotation or config) and scans all declared services to see if they use one of the traits, then appropriately add a lazy call the setter method.
3. then when the service is called the service required by the trait is injected. 
# Advantages
1. removes boilerplate in services, or many calls to setter methods.
2. defines the dependency in the file, thus keeping the services container clean.
3. provides a pure php alternative to the Expression Syntax for manipulating injected services.
4. done in compiler passes, so mostly efficient. 
5. keeps auto complete in IDE's without the need for a plugin. 
# Disadvantages
1. Service dependencies are no longer only explicitly defines in the services yaml. 
   - However since routes can alternatively be setup inside the controller, I don't see why is a problem. its merely the same type of alternative. Within the file where its relevant or gathered in one location. 
2. requires PHP 5.4 

Comments? (This is really just satisfying a curiousity)

**Comment by unkind** [comment]
There was similar discussion: #12421. In practice, there are almost no advantages in terms of lines of code/configs. Also traits promote bad practices (in context of DI?).

**Comment by WiR3D-kNiGhT** [comment]
Bad practice being the setter injection? Yes can't argue that, because it needs a publically available setter. 

I read #12421 and it is similar but not the same, but some of the points are valid. It mainly focusses on EventDispatcher which is a fairly different use case to what I am suggesting. 

Humour me though. Why is setter injection a bad practice? Because it allows devs to change the context of the object and can introduce bugs that are hard to debug? 

which I agree with. fair enough.

**Comment by stof** [comment]
Several issues with your proposal:

- traits don't define a contract at all. Knowing that a class uses a trait does not give you any guarantee about the methods available in it (you can rename methods when using a trait), so your feature is broken by design. Btw, this is also the reason why PHP does not have a ``class_uses`` function while it has ``class_implements``
- contrary to what you reports in your advantages, this does not save setter calls at all. Even if the setter is defined in the trait rather than in the class, it still needs to be called
- what is more efficient about being done in the compiler pass ? Compiler passes can only manipulate service definitions, so it won't be faster than defining the definition explicitly
- this does not replace the expression syntax at all. You are confusing different things here
- there is nothing like lazy setter calls in the DI component
- automatic injection based on implemented interfaces (which define a contract while traits do not and so would be cleaner) was implemented in early dev versions of Symfony 2.0. We removed it because it created lots of issues compared to the benefit (no way to have several different loggers for instance, or to avoid setting a logger in a service, at least not without a complex syntax to disable the interface-based injection)
- the IDE autocompletion has nothing to do here. Classes using DI will have proper autocompletion already as they get their dependencies injected in them. The case where you need an IDE plugin is when using service location, not when using DI. And traits cannot replace service location (they can only provide service location shortcuts, and Symfony does not need to be changed for that)

> Humour me though. Why is setter injection a bad practice? Because it allows devs to change the context of the object and can introduce bugs that are hard to debug?
> 
> which I agree with. fair enough.

The issue with setter injections is that it makes your dependency an optional mutable one. This means 2 reasons to introduce bugs:

- your class needs to consider the dependency as optional (there is not guarantee that the setter will be called), forcing to add boilerplate code for any access to this dependency
- your class needs to account for the possibility to call the getter again at any time, making your service much harder to reason about.

**Comment by unkind** [comment]
> Why is setter injection a bad practice?

The main issue is that object may be in an invalid state, i.e. I can create object and can't be sure if it  works, maybe it uses `$this->logger` directly and fails with Fatal error. It forces to add boilerplate code as @stof said above. Also it's harder to see class dependencies and it makes service mutable as you said. Constructor is the most natural and clear way to describe dependencies.

**Comment by WiR3D-kNiGhT** [comment]
What an answer. Thanks.

Learnt a few things there.  

> PHP does not have a class_uses 

I assume this is a typo? 

> * contrary to what you reports in your advantages, this does not save setter calls at all. Even if the setter is defined in the trait rather than in the class, it still needs to be called
what is more efficient about being done in the compiler pass ? 
* Compiler passes can only manipulate service definitions, so it won't be faster than defining the definition explicitly
* this does not replace the expression syntax at all. 
* You are confusing different things here
there is nothing like lazy setter calls in the DI component

Also I think there was some poor communication on my part, I was trying to express that it doesn't have any overhead over the current method of defining it in the services yaml. 


I was aware of this:
> automatic injection based on implemented interfaces (which define a contract while traits do not and so would be cleaner) was implemented in early dev versions of Symfony 2.0. We removed it because it created lots of issues compared to the benefit (no way to have several different loggers for instance, or to avoid setting a logger in a service, at least not without a complex syntax to disable the interface-based injection)

But you raise an important point I did not consider, in overriding. 

The IDE autocompletion was just a tidbit. could easily not have mentioned it. 

## In conclusion: 
In the world of lesser evils it makes sense to have more repetition or boilerplate with tags in the service definition then to hope that a method exists on an object just because it uses a trait, and have the crap side effect that is mutable dependencies. 

Thanks for satisfying my curiosity.

**Comment by KingCrunch** [comment]
Just want to correct something

@stof 

> traits don't define a contract at all. Knowing that a class uses a trait does not give you any guarantee about the methods available in it (you can rename methods when using a trait), so your feature is broken by design.

No, it's an alias. The method is still available with the "old" name.

**Comment by rijnhard** [comment]
I'm going to reopoen this. because things have changed quite a bit since I logged this issue and stumbled on this again.

If we have `_instanceof` in DI that works with interfaces and allows us to set default configuration based on an Interface then why not the same for traits? I would argue that traits are more utilitarian.

**Comment by rijnhard** [comment]
I know of the @required annotation but I feel like thats an anti-pattern (purely my own opinion) because I like to see where dependencies are set in as few locations as possible e.g. the services.yaml file

**Comment by ro0NL** [comment]
@rijnhard i believe

> traits don't define a contract at all. Knowing that a class uses a trait does not give you any guarantee about the methods available in it 

still applies isnt it :)

**Comment by jvasseur** [comment]
@rijnhard `_instanceof` is related to the `instanceof` PHP operator which doesn't work for traits so my opinion would be that nothing changed here.

**Comment by stof** [comment]
I vote -1 against that proposal, for the reason highlighted by @jvasseur just above.

**Comment by javiereguiluz** [comment]
Let's close this issue for the reasons given above. In any case, @rijnhard thanks for taking the time to send proposals to improve Symfony!
